### ISSUE TITLE
Tapping compass rose on waypoint tab triggers scrolling instead

### ISSUE THREAD

**Issue by andrcoding** [cgeo/cgeo#9992]
Since an update last summer/autumn, it has become difficult to access the map view directly by long pressing on the compass (icon on the right) of a waypoint (Opened cache-most left page - waypoints). Instead of pressing the compass, the scroll bar is often triggered. Before everything works fine. Moving the compass a little bit left or changing the touching areas would be helpful.

![cgeo](https://user-images.githubusercontent.com/78544594/107421389-2fc9ad00-6b1a-11eb-8173-7e72327fa106.png)

Edit: One of the system I have tested it: [system information.txt](https://github.com/cgeo/cgeo/files/5954264/system.information.txt)

**Comment by fm-sys** [comment]
Can you please post which c:geo and Android version you use, or ever better post the system information, which you will find in c:geo Menu -> About c:geo -> Swipe right to System?

At least for me, the problem does not exist, so it's either meanwhile fixed or a device specific problem...

**Comment by Narrat** [comment]
Just a FYI post, as I noticed that the requested information got edited into the first post :)

**Comment by fm-sys** [comment]
Thanks for the hint. 

Seems like the cgeo version is up-to-date. So there might be a bug in our "fastscroll magic" which should normally gets deactivated after one second with no scrolling. 

@andrcoding does this happen to you always or only after performing a fast fling?

**Comment by andrcoding** [comment]
From my experience, it happens not only after performing a fast fling. However, it seems that the "fastscroll magic" feature is not deactivated all the time after one second. Typically it is about two seconds. I will watch this behavior the next times of usage to make a clearer statement if "fastscroll magic" is deactivated while the problem occur. I will give an update.
In general, I have first noticed it after a major update but I am not sure which one it was. It could be version 2020.08.04.

@Narrat Thanks for informing. I thought there will be a notice.

**Comment by andrcoding** [comment]
After some testing some feedback: It seems that the "fastscroll magic" feature is not always getting activated which is ok. In general, a reduction of the deactivation time will be good as it sometimes takes 2-3 seconds. In my opinion, 0.5 seconds should be totally enough. Thus, my problem is only related to time.

**Comment by Lineflyer** [comment]
If this is configurable in the code we could indeed reduce it, but I suggest something around 1 second to allow users to view the scroll position before it disappears.

**Comment by Lineflyer** [comment]
I stumbled over this issue while caching out in the field lately.
I am not sure why it affected me never before, however this time it was really annoying not to be able to activate waypoint navigation after scrolling.

I really think we should avoid that overlap between the clickable icon and the fast scroll indicator (or make it disappear really fast after scrolling).

**Comment by murggel** [comment]
> I stumbled over this issue while caching out in the field lately.
> I am not sure why it affected me never before, however this time it was really annoying not to be able to activate waypoint navigation after scrolling.
> 
> I really think we should avoid that overlap between the clickable icon and the fast scroll indicator (or make it disappear really fast after scrolling).

Maybe this came for you now with the new svg, I guess I reduced the button size  - was not aware of the scroll indicator. 
Maybe the button can be moved more to the left?

**Comment by Lineflyer** [comment]
> Maybe the button can be moved more to the left?

That would be maybe the easiest way to mitigate the problem. I guess its just a matter of a few pixels thus it should not steal to much space from the title.

**Comment by fm-sys** [comment]
The situation hasn't changed with migration to SVG. It's the same as it was before. 

But do we really need fastscroll for waypoints at all? Disabling fastscroll totally would be an even easier solution...

**Comment by Lineflyer** [comment]
> But do we really need fastscroll for waypoints at all? Disabling fastscroll totally would be an even easier solution...

AFAIR we had this discussion and finally decided to implement it as the original feature request was about caches with large amount of waypoints, where fast-scroll is helpful. Additionally we still have the issue, that after editing a waypoint the dialog jumps to the top (which again requires scrolling) and this was a "workaround".

Thats why I said maybe the easiest is to move the icon instead of removing a feature.

**Comment by Lineflyer** [comment]
Or doublechecking the timeout for the fastscroll bar to make it disappear faster?

**Comment by Lineflyer** [comment]
Thanks @moving-bits for adding some extra padding, but somehow it has not improved for me.
I might need to look on the emulator after my vacation but it seems, that even if I touch outside the scrollbar it still takes it as an input for the scrollbar instead of the navigation icon.

**Comment by moving-bits** [comment]
Hm, I cannot reproduce that behavior in the emulator at all, button hitting the scroll handle (or the icon) with a mouse is way easier probably as with my finger. Will try on a "real mobile" the next days (weekend maybe).

**Comment by Lineflyer** [comment]
I recently did not go out for caching for a cache with many waypoints, thus I cannot judge right now.
Need to test in the nightlies next days.

Please lets keep this open.

**Comment by andrcoding** [comment]
I have tested the current version. In my opinion, it has not changed the situation at all. I even assume that it has got a bit worse.

**Comment by Lineflyer** [comment]
Gave a better title to the issue:

I again stumbled over this issue in conjunction with https://github.com/cgeo/cgeo/issues/1905#issuecomment-1001155124

- Open GC9JBGX which has a lot of waypoints
- While working on the solution you have to scroll through the waypoints and open the map for each of them
- It happend very often, that trying to open the map by tapping on the compass rose I instead triggered the fast scroll

Is it possible to add some more padding and/or make the fast scroll bar disappear sooner?

**Comment by moving-bits** [comment]
Hm, seems to be device- (or vendor-) specific. I'm still not able to reproduce the problem. Even with fastscroll bar still active, tapping on compass rose opens the compass.
But looking at the source, the additional margins introduced some time ago somehow got removed with ec2ccdd, so let's create a PR to re-add them.

**Comment by Lineflyer** [comment]
That could indeed be the reason, because I was really wondering to get into this massive problem again after a while.

I guess it could be kind of device specific as my S20 as rounded edges.

**Comment by ztNFny** [comment]
I had already added padding to the compass button in ec2ccdd1e1dbb01449829c07a36b4174946a5483, it's click target is a lot larger than the icon itself.

**Comment by ztNFny** [comment]
The clickable area of the fastscroll handler is huge, I measured it with some trial and error:
![image](https://user-images.githubusercontent.com/1258173/147592967-3c6930e7-df1d-478b-9956-d5dd1627ba51.png)
So the area extends far beyond the button! If unless we leave the right 10% of the screen blank then this can't be fixed by padding/margin.

**Comment by Lineflyer** [comment]
This would mean its just not possible to click the icon while fastscroll is active. :(
Can we alter the clickable area of the fastscroll handler?

Interesting enough it seemed to work better until recently (maybe before ec2ccdd) but I am not sure though.

**Comment by moving-bits** [comment]
Do you have a chance to test PR #12420 to see if "the old way" works better in your case?

**Comment by ztNFny** [comment]
> Do you have a chance to test PR #12420 to see if "the old way" works better in your case?

Have you looked at the screenshots in https://github.com/cgeo/cgeo/issues/9992#issuecomment-1002107786 ? And still expect the old to be easier to click?
#12420 doesnt restore the old btw, it adds margin to the existing padding, moving the icon even further to the left.

**Comment by moving-bits** [comment]
> Have you looked at the screenshots in [#9992 (comment)](https://github.com/cgeo/cgeo/issues/9992#issuecomment-1002107786) ?

Yes, have seen them. But according to @Lineflyer's experience it has worked on his device for some time (after adding extra padding to the button), and does no longer work lately, probably after the changes introduced with ec2ccdd - which is what I want to either verify or falsify. Android's layout system is not always intuitive.

> And still expect the old to be easier to click? #12420 doesnt restore the old btw, it adds margin to the existing padding, moving the icon even further to the left.

Which might (or might not) be something to mitigate the problem @Lineflyer experiences.

**Comment by ztNFny** [comment]
> Which might (or might not) be something to mitigate the problem @Lineflyer experiences.

If you think ec2ccdd caused some issue then it'd make sense to revert all the changes to that file.
Right now you're adding 13dp margin to an image that already has 21dp padding around it (meaning 10.5dp on each side), pushing the icon to the left by 23.5dp.
The fastscroll area is 46dp wide, so with your 13dp margin right and an icon size of 45dp the left 12dp of the icon would be clickable - but it'll now be easy to miss the button if you click on the right side of the screen - as margin is not part of the clickable area, unlike the padding.

Unless you want to move the icon out of the area of the fastscroll handler - which is approximately 46dp wide, so a 46dp margin-right would have to be added instead of that 13dp margin - this can't be changed with icon placement.

What can be done however is a) disabling fastscroll for that tab or b) change the timeout in FastScrollListener.java

**Comment by Lineflyer** [comment]
Maybe we should take the easy approach and test = Reduce the timeout to e.g. 500ms.

**Comment by Lineflyer** [comment]
> Do you have a chance to test PR #12420 to see if "the old way" works better in your case?

Unfortunately no. I checked current master on an emulator, but here everything is fine and I can currently not install a debug build on my main device (the S20 with the rounded edges).

**Comment by Lineflyer** [comment]
One important thing I just noticed:
Short click on compass rose seems less problematic on my device (works mostly also if fastscroll bar is still shown) but long click (to open the map for the waypoint) does maybe work in 5% of cases only.

**Comment by ztNFny** [comment]
No issued with long-clicks here. Anything else?

**Comment by Lineflyer** [comment]
On my tablet it happens also nearly always. Tried to make a video with developer tool for visual tap/scroll/fling  active.

https://user-images.githubusercontent.com/949669/147786717-c42d03ec-e9fe-4ef3-872a-4ae5f4064d3f.mp4

**Comment by ztNFny** [comment]
Of course it happens when the fastscroll mode is active, that is how that mode is supposed to work. If you get the big scrollbar the clicking somewhere brings you there.
The video only shows expected behaviour.

**Comment by Lineflyer** [comment]
Found this one, which is old but pretty much confirms all findings:
https://stackoverflow.com/questions/33619453/scrollbar-touch-area-in-android-6

First workaround is what we already do: Disable/Enable on demand (otherwise the problem would be much worse)
The second workaround sounds more interesting maybe?? What do you think?

**Comment by ztNFny** [comment]
> When you get the big scrollbar then clicking somewhere in the 46dp wide area reserved by fast scroll scrolls.

Actually the "click" in there is wrong. You have to drag, even by just a few px, for fast scroll to jump to that position. So if your long-click in fast-scroll mode stays perfectly still it won't jump to that position.
But again, this is how fast scroll is supposed to work. If you don't want fact scroll then wait for it to time out (1s)

**Comment by ztNFny** [comment]
> First workaround is what we already do: Disable/Enable on demand (otherwise the problem would be much worse) The second workaround sounds more interesting maybe?? What do you think?

IMHO we shouldn't interfer with a standard accessibility feature of the platform. Just wait the 1s

**Comment by Lineflyer** [comment]
Or decrease to e.g. 500ms or less to lower the probability of this to happen but still achieve fast scrolling on long lists?

**Comment by Lineflyer** [comment]
> Of course it happens when the fastscroll mode is active,
I am still wondering why it does not happen on an emulated device, therefore I posted the video.

I will play around with the timeout value to see how low it can be set to still allow good interaction but decrease the probability to block the touches.
